src/utils/PluginManager.ts

```diff
@@ -0,0 +1,123 @@
+import path from 'path';
+import fs from 'fs-extra';
+import { execFile } from 'child_process';
+import util from 'util';
+import { InvalidHookError, MissingPluginError, QuestionableNpmNameError } from '../errors';
+import { PluginConfiguration } from '../fshtypes';
+import { logger } from './FSHLogger';
+
+export const AVAILABLE_HOOKS = [
+  'beforeFillTank',
+  'beforeLoadCustomResources',
+  'afterLoadCustomResources',
+  'afterExportFHIR'
+];
+
+export class PluginManager {
+  static hooks = new Map<string, ((...args: any) => any)[]>();
+
+  static async processPluginConfiguration(
+    plugins: PluginConfiguration[],
+    pluginBasePath: string
+  ): Promise<void> {
+    for (const plugin of plugins) {
+      // if the plugin name is given as name@version, it's something we may need to load from npm
+      // otherwise, we load it from the filesystem
+      // when we load from the filesystem, it might be at pluginBasePath, or at pluginBasePath/node_modules
+      const versionSeparator = plugin.name.lastIndexOf('@');
+      let pluginName: string;
+      let pluginVersion: string;
+      if (versionSeparator > 0) {
+        pluginName = plugin.name.slice(0, versionSeparator);
+        pluginVersion = plugin.name.slice(versionSeparator + 1);
+      } else {
+        pluginName = plugin.name;
+      }
+      try {
+        if (pluginVersion != null) {
+          // this is a plugin that can be acquired from npm.
+          // it might already be installed, so check for that first.
+          const existingInstallationPath = path.join(pluginBasePath, 'node_modules', pluginName);
+          if (fs.existsSync(path.join(existingInstallationPath, 'package.json'))) {
+            // check the version number in package.json
+            const packageContents = fs.readJsonSync(
+              path.join(existingInstallationPath, 'package.json')
+            );
+            if (packageContents.version === pluginVersion) {
+              PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+            } else {
+              await PluginManager.installFromNpm(pluginBasePath, plugin.name);
+              PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+            }
+          } else {
+            await PluginManager.installFromNpm(pluginBasePath, plugin.name);
+            PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+          }
```

Comment:
I suppose it could be simplified even further if you just want the block above to only handle doing the NPM install.  Then you could take the block below (lines 57 - 66) out of the else block and let them load the npm module you just installed.  The main difference would be that a project could override the npm installed version w/ a local version instead.  Which maybe is a good thing?  I'm not sure.